### PR TITLE
fix(build): handle dynamic requests in JS files with JSX

### PR DESCRIPTION
**Pull request by NathanDrake2406**
## Summary
- narrow this PR to the remaining dynamic-request gap not covered by #2253
- parse plain `.js` app modules as JSX before the dynamic-request pre-transform walks them
- add a focused build regression for `.js` App Router files with JSX and guarded fully dynamic requests

## Why

The broader dynamic-request analysis work from this PR was superseded by #2253. Current `main` still missed one edge case: a `.js` page/layout containing JSX and guarded `require(dynamic)` / `import(dynamic)` could fail before the pre-transform ran, leaving `vite-plugin-commonjs` to throw on `require(dynamic)`.

## Validation
- `vp test run tests/dynamic-requests-build.test.ts`
- `vp check --fix packages/vinext/src/plugins/ignore-dynamic-requests.ts tests/dynamic-requests-build.test.ts`
- commit hook: staged checks, full check, staged tests, and knip

### CODE REVIEW

**Comment by NathanDrake2406** [comment]
All three blocking issues are valid. Implemented fixes and pushed.

**Blocking issue 1 — `.js` with JSX:** Valid. The plugin runs before `vinext:jsx-in-js`, so `.js` files containing JSX would fail to parse with `lang: "js"` and silently fall through. Fixed by returning `lang: "jsx"` for `.js`/`.mjs`/`.cjs` extensions in `parserLangForId`. Added integration test that builds an `app/page.js` with JSX + guarded dynamic requests.

**Blocking issue 2 — name-based `require` matching:** Valid. Rewriting every call whose callee is named `require` changes program semantics when the user has a local binding. Fixed by adding `walkAstWithBindings`, which tracks lexical scopes (function params, declarations, imports, catch bindings, destructuring patterns). `isVeryDynamicRequireCall` now checks `isRequireBound()` and skips rewriting when `require` is shadowed. Added unit tests for local `const require`, function parameter `require`, imported `require`, and function-named-`require`.

**Blocking issue 3 — TS wrapper expressions hiding static parts:** Valid. `TSAsExpression`, `TSTypeAssertion`, `TSNonNullExpression`, `TSInstantiationExpression`, `ParenthesizedExpression`, and `ChainExpression` were treated as unknown nodes, causing misclassification. Fixed by adding `unwrapTransparentExpression` that recurses through these wrappers before `requestHasStaticPart` evaluates the expression. Added unit tests for `require(('./' + name) as string)`, `` import((`./${name}`) as string) ``, and `require(path!)`.

All new and existing tests in `tests/build-optimization.test.ts` and `tests/cjs.test.ts` pass. `vp check` is clean.

**Comment by NathanDrake2406** [comment]
Addressed the latest REQUEST_CHANGES review and pushed `5b3e9406`.

Assessment: all three comments were valid.

1. `var require` inside `catch` bodies: valid. The hoist pre-pass now recurses through `CatchClause`, so `var require` in a catch body binds the enclosing hoisting scope. Added regressions for both the in-catch call and the post-catch call.

2. `import type` runtime binding: valid. Runtime scope collection now skips `ImportDeclaration` nodes with `importKind === "type"` and individual specifiers with `importKind === "type"`, while preserving value imports. Added regressions for full type-only imports and mixed type/value imports.

3. `satisfies` transparent wrapper: valid. `TSSatisfiesExpression` now unwraps through the same path as `as`, non-null, parenthesized, and instantiation wrappers. Added regressions for both `require` and dynamic `import` with static path parts.

Also merged current `upstream/main` to use the new shared `ast-utils` and replaced the plugin-local AST range/child/binding helpers with that utility.

Verification:
- `vp test run tests/build-optimization.test.ts -t "catch-body|import type|type-only import specifier|satisfies"` passed: 6 tests.
- `vp check tests/build-optimization.test.ts packages/vinext/src/plugins/ignore-dynamic-requests.ts` passed: format, lint, type checks clean.
- `vp test run tests/build-optimization.test.ts` passed: 123 tests.

**Comment by NathanDrake2406** [comment]
Addressed the latest REQUEST_CHANGES review and pushed `41e776f4`.

Assessment: both comments were valid.

1. `var require` under a labelled statement: valid. Labels do not introduce a new scope, but the var-hoist pre-pass was not recursing through `LabeledStatement`. Added it to the hoist recursion set and added regressions for the in-labelled-block call and the post-labelled-block call.

2. Empty string literals: valid. `require("")` and `import("")` are statically known literal requests, not very dynamic requests. `requestHasStaticPart()` now treats any string literal as static while leaving the non-empty template-quasi boundary unchanged. Added regressions for both forms.

Verification:
- Red check before fix: `vp test run tests/build-optimization.test.ts -t "labelled-block|static literal request"` failed on the four new regressions.
- After fix: same command passed, 4 tests.
- `vp test run tests/build-optimization.test.ts` passed, 127 tests.
- `vp check tests/build-optimization.test.ts packages/vinext/src/plugins/ignore-dynamic-requests.ts` passed: format, lint, and type checks clean.

**Comment by NathanDrake2406** [comment]
@codex adversarial review

**Comment by NathanDrake2406** [comment]
@codex review

File: packages/vinext/src/plugins/ignore-dynamic-requests.ts [hunk 1]
```diff
@@ -0,0 +1,619 @@
...
+}
+
+export function createIgnoreDynamicRequestsPlugin(): Plugin {
+  return {
+    name: "vinext:ignore-dynamic-requests",
+    enforce: "pre",
+    transform(code, id) {
+      if (!isTransformableModuleId(id)) return null;
+      return ignoreVeryDynamicRequests(code, id);
+    },
```
[the remark this answers is not in the document]
Comment (NathanDrake2406) [reply packages/vinext/src/plugins/ignore-dynamic-requests.ts:911]: Added the dev/build parity comment in `a3f506f9` above the transform hook. The plugin remains intentionally unguarded by `apply: "build"` because the CommonJS analyzer path it protects can run in both modes.

File: packages/vinext/src/plugins/ignore-dynamic-requests.ts [hunk 2]
```diff
@@ -0,0 +1,619 @@
...
+    if (isVeryDynamicImportExpression(node)) {
+      overwriteNode(node, dynamicImportFailureExpression());
+    }
+  });
+
+  if (!changed) return null;
+
+  return {
+    code: output.toString(),
+    map: output.generateMap({ hires: true, source: id }),
```
[the remark this answers is not in the document]
Comment (NathanDrake2406) [reply packages/vinext/src/plugins/ignore-dynamic-requests.ts]: Fixed in `a3f506f9`. `ignoreVeryDynamicRequests()` now computes `source = cleanModuleId(id)` once, uses it for `parseAst`, and passes the same value to `generateMap`. Added a source-map regression for query/hash module IDs.

File: packages/vinext/src/plugins/ignore-dynamic-requests.ts [hunk 3]
```diff
@@ -0,0 +1,619 @@
...
+        collectBindingNames(node.id, scopeStack[scopeStack.length - 1]);
+      }
+      const body = toAstRecord(node.body);
+      if (body) walkNode(body);
+      if (node.id) scopeStack.pop();
+      return;
+    }
+
+    // Generic recursion for non-scope nodes
+    forEachAstChild(node, walkNode);
```
[the remark this answers is not in the document]
Comment (NathanDrake2406) [reply packages/vinext/src/plugins/ignore-dynamic-requests.ts]: Leaving this unchanged intentionally. The transform targets modules parsed as ESM/strict-mode-compatible source, where `with` is a syntax/runtime non-starter before this compatibility branch matters. Adding dynamic-object binding semantics for `with` would complicate the walker for a path that cannot validly appear in the module graph this plugin is meant to handle.

File: packages/vinext/src/plugins/ignore-dynamic-requests.ts [hunk 4]
```diff
@@ -0,0 +1,619 @@
...
+  return nodeArray(value).flatMap((entry) => {
+    const node = toAstRecord(entry);
+    return node ? [node] : [];
+  });
+}
+
+function singleArgument(node: AstRecord): AstRecord | null {
+  const args = astArray(node.arguments);
+  return args.length === 1 ? (args[0] ?? null) : null;
+}
```
[the remark this answers is not in the document]
Comment (NathanDrake2406) [reply packages/vinext/src/plugins/ignore-dynamic-requests.ts:80]: Fixed in `a3f506f9`. `singleArgument()` now treats a sole `SpreadElement` as no positional request argument, so `require(...args)` falls through to the existing analyzer/runtime path. Added a regression for `require(...args)`.

File: packages/vinext/src/plugins/ignore-dynamic-requests.ts [hunk 5]
```diff
@@ -0,0 +1,625 @@
...
+
+  const request = singleArgument(node);
+  if (!request) return false;
+
+  return !requestHasStaticPart(request);
+}
+
+function isVeryDynamicImportExpression(node: AstRecord): boolean {
+  if (node.type !== "ImportExpression") return false;
+  return !requestHasStaticPart(toAstRecord(node.source));
```
[the remark this answers is not in the document]
Comment (NathanDrake2406) [reply packages/vinext/src/plugins/ignore-dynamic-requests.ts]: Leaving this unchanged intentionally. For the ignored very-dynamic path, the contract is to replace the whole request expression with the deterministic runtime failure expression. That matches the Turbopack-style framing: no graph edge is created and execution fails at that expression rather than evaluating subexpressions that are part of an unanalyzable request. Preserving option/request side effects would make this branch less like a direct expression replacement and would add complexity for a path whose observable result is failure.

File: packages/vinext/src/plugins/ignore-dynamic-requests.ts [hunk 6]
```diff
@@ -0,0 +1,625 @@
...
+      resolvedId === VINEXT_SOURCE_ROOT ||
+      resolvedId.startsWith(`${VINEXT_SOURCE_ROOT}${path.sep}`)
+    ) {
+      return false;
+    }
+  }
+  return MODULE_EXTENSIONS.has(path.extname(cleanId));
+}
+
+export function mayContainVeryDynamicRequestTarget(code: string): boolean {
```
[the remark this answers is not in the document]
Comment (NathanDrake2406) [reply packages/vinext/src/plugins/ignore-dynamic-requests.ts]: Leaving this as-is. The regex is only a cheap precheck to avoid parsing ordinary static-import modules; false positives in comments/strings pay one parse and then the AST walk returns `null`. Narrowing this further would either become a lexer-lite path or risk missing real call-form targets, so the current cost/correctness boundary is intentional.

File: packages/vinext/src/index.ts [hunk 7]
```diff
@@ -1024,6 +1025,10 @@ export default function vinext(options: VinextOptions = {}): PluginOption[] {
     ...(viteMajorVersion >= 8 ? [] : [tsconfigPaths()]),
     // React Fast Refresh + JSX transform for client components.
     reactPluginPromise,
+    // Next.js/Turbopack ignores "very dynamic" requests with no static path
+    // part during graph analysis. Preserve that behaviour before the CJS
+    // plugin tries to expand require(dynamic) as a static glob.
+    createIgnoreDynamicRequestsPlugin(),
     // Transform CJS require()/module.exports to ESM before other plugins
     // analyze imports (RSC directive scanning, shim resolution, etc.)
     commonjs(),
```
Comment (james-elicx) [inline packages/vinext/src/index.ts]: tbh this feels more like a hack reading the comment

Comment (NathanDrake2406) [reply packages/vinext/src/index.ts]: Valid. Addressed in `1eb70809` by rewriting the placement comment around the actual graph-construction boundary instead of the downstream CommonJS symptom.

I also applied the small internal vocabulary recommendation in `f4bf5aae`: the transform now classifies request expressions as `static`, `partly-static`, or `very-dynamic`, and the rewrite path explicitly only removes `very-dynamic` requests from graph analysis.

I intentionally did not split the Vite adapter or move the scope walker in this PR. The plugin adapter is already thin at the public boundary, and the scope walker still has no second caller; extracting either now would add churn beyond the feedback’s own recommended move. Verification: `vp check packages/vinext/src/index.ts`, `vp check --fix packages/vinext/src/plugins/ignore-dynamic-requests.ts tests/build-optimization.test.ts`, and `vp test run tests/build-optimization.test.ts` all passed.

**Comment by NathanDrake2406** [comment]
Reworked this PR down to the remaining useful part.

#2253 superseded the broad dynamic-request analysis changes, so I force-updated this branch to a single commit on current `main`. The PR now only fixes the `.js` + JSX edge case where the dynamic-request pre-transform parsed the module as plain JS, fell through, and left `vite-plugin-commonjs` to fail on guarded `require(dynamic)`.

Validation:
- `vp test run tests/dynamic-requests-build.test.ts`
- `vp check --fix packages/vinext/src/plugins/ignore-dynamic-requests.ts tests/dynamic-requests-build.test.ts`
- commit hook passed staged checks, full check, staged tests, and knip
